File: herald-admin-typescript/src/namespaces/messages.ts
import type { HttpTransport } from "../transport.js";
import type { EventList, EventPublishResult, ReactionSummary } from "../types.js";

export class EventNamespace {
  constructor(private transport: HttpTransport) {}

  async publish(
    streamId: string,
    sender: string,
    body: string,
    options?: { meta?: unknown; parentId?: string; excludeConnection?: string },
  ): Promise<EventPublishResult> {
    const headers: Record<string, string> = {};
    if (options?.excludeConnection) {
      headers["X-Exclude-Connection"] = options.excludeConnection;
    }
    return this.transport.request<EventPublishResult>(
      "POST",
      `/streams/${encodeURIComponent(streamId)}/events`,
      { sender, body, meta: options?.meta, parent_id: options?.parentId },
      headers,
    );
  }

  async list(
    streamId: string,
    options?: { before?: number; after?: number; limit?: number; thread?: string },
  ): Promise<EventList> {
    const params = new URLSearchParams();
    if (options?.before !== undefined) params.set("before", String(options.before));
    if (options?.after !== undefined) params.set("after", String(options.after));
    if (options?.limit !== undefined) params.set("limit", String(options.limit));
    if (options?.thread !== undefined) params.set("thread", options.thread);
    const qs = params.toString();
    const path = `/streams/${encodeURIComponent(streamId)}/events${qs ? `?${qs}` : ""}`;
    return this.transport.request<EventList>("GET", path);
  }

  /** @chat Chat-specific operation — event deletion. */
  async delete(streamId: string, eventId: string): Promise<void> {
    await this.transport.request(
      "DELETE",
      `/streams/${encodeURIComponent(streamId)}/events/${encodeURIComponent(eventId)}`,
    );
  }

  /** @chat Chat-specific operation — event editing. */
  async edit(streamId: string, eventId: string, body: string): Promise<void> {
    await this.transport.request(
      "PATCH",
      `/streams/${encodeURIComponent(streamId)}/events/${encodeURIComponent(eventId)}`,
      { body },
    );
  }

  /** @chat Chat-specific operation — reaction queries. */
  async getReactions(streamId: string, eventId: string): Promise<ReactionSummary[]> {
    const res = await this.transport.request<{ reactions: ReactionSummary[] }>(
      "GET",
      `/streams/${encodeURIComponent(streamId)}/events/${encodeURIComponent(eventId)}/reactions`,
    );
    return res.reactions;
  }

  async trigger(streamId: string, event: string, data?: unknown, excludeConnection?: number): Promise<void> {
    const body: Record<string, unknown> = { event };
    if (data !== undefined) body.data = data;
    if (excludeConnection !== undefined) body.exclude_connection = excludeConnection;
    await this.transport.request("POST", `/streams/${encodeURIComponent(streamId)}/trigger`, body);
  }

  async search(
    streamId: string,
    query: string,
    limit?: number,
  ): Promise<EventList> {
    const params = new URLSearchParams({ q: query });
    if (limit !== undefined) params.set("limit", String(limit));
    const path = `/streams/${encodeURIComponent(streamId)}/events/search?${params}`;
    return this.transport.request<EventList>("GET", path);
  }
}

File: herald-admin-ruby/lib/herald_admin/messages.rb
# frozen_string_literal: true

module HeraldAdmin
  class MessageNamespace
    def initialize(transport)
      @t = transport
    end

    def send(room_id, sender:, body:, meta: nil, parent_id: nil, exclude_connection: nil)
      req = { sender: sender, body: body }
      req[:meta] = meta if meta
      req[:parent_id] = parent_id if parent_id
      req[:exclude_connection] = exclude_connection if exclude_connection
      data = @t.request("POST", "/rooms/#{esc(room_id)}/messages", req)
      MessageSendResult.new(id: data["id"], seq: data["seq"], sent_at: data["sent_at"])
    end

    def list(room_id, before: nil, after: nil, limit: nil, thread: nil)
      params = []
      params << "before=#{before}" if before
      params << "after=#{after}" if after
      params << "limit=#{limit}" if limit
      params << "thread=#{ERB::Util.url_encode(thread)}" if thread
      qs = params.empty? ? "" : "?#{params.join("&")}"
      data = @t.request("GET", "/rooms/#{esc(room_id)}/messages#{qs}")
      messages = data["messages"].map do |m|
        Message.new(id: m["id"], room: m["room"] || room_id, seq: m["seq"],
                    sender: m["sender"], body: m["body"], meta: m["meta"],
                    parent_id: m["parent_id"], edited_at: m["edited_at"], sent_at: m["sent_at"])
      end
      MessageList.new(messages: messages, has_more: data["has_more"] || false)
    end

    def delete(room_id, message_id)
      @t.request("DELETE", "/rooms/#{esc(room_id)}/messages/#{esc(message_id)}")
    end

    def edit(room_id, message_id, body)
      @t.request("PATCH", "/rooms/#{esc(room_id)}/messages/#{esc(message_id)}", { body: body })
    end

    def get_reactions(room_id, message_id)
      data = @t.request("GET", "/rooms/#{esc(room_id)}/messages/#{esc(message_id)}/reactions")
      data["reactions"]
    end

    def trigger(room_id, event, data: nil, exclude_connection: nil)
      body = { event: event }
      body[:data] = data if data
      body[:exclude_connection] = exclude_connection if exclude_connection
      @t.request("POST", "/rooms/#{esc(room_id)}/trigger", body)
    end

    def search(room_id, query:, limit: nil)
      params = ["q=#{ERB::Util.url_encode(query)}"]
      params << "limit=#{limit}" if limit
      qs = params.join("&")
      data = @t.request("GET", "/rooms/#{esc(room_id)}/messages/search?#{qs}")
      messages = data["messages"].map do |m|
        Message.new(id: m["id"], room: m["room"] || room_id, seq: m["seq"],
                    sender: m["sender"], body: m["body"], meta: m["meta"],
                    parent_id: m["parent_id"], edited_at: m["edited_at"], sent_at: m["sent_at"])
      end
      MessageList.new(messages: messages, has_more: data["has_more"] || false)
    end

    private

    def esc(s) = ERB::Util.url_encode(s)
  end
end
